airbyte_cdk/sources/declarative/decoders/composite_decoder.py

```diff
@@ -0,0 +1,98 @@
+import gzip
+import json
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from io import BufferedIOBase
+from typing import Any, Generator, MutableMapping, Optional
+
+import pandas as pd
+import requests
+from numpy import nan
+
+from airbyte_cdk.sources.declarative.decoders.decoder import Decoder
+
+logger = logging.getLogger("airbyte")
+
+
+@dataclass
+class Parser(ABC):
+    @abstractmethod
+    def parse(
+        self, data: BufferedIOBase, *args, **kwargs
+    ) -> Generator[MutableMapping[str, Any], None, None]:
+        """
+        Parse data and yield dictionaries.
+        """
+        pass
+
+
+@dataclass
+class GzipParser(Parser):
+    inner_parser: Parser
+
+    def parse(
+        self, data: BufferedIOBase, *args, **kwargs
+    ) -> Generator[MutableMapping[str, Any], None, None]:
+        """
+        Decompress gzipped bytes and pass decompressed data to the inner parser.
+        """
+        gzipobj = gzip.GzipFile(fileobj=data, mode="rb")
+        yield from self.inner_parser.parse(gzipobj)
+
+
+@dataclass
+class JsonLineParser(Parser):
+    encoding: Optional[str] = "utf-8"
+
+    def parse(
+        self, data: BufferedIOBase, *args, **kwargs
+    ) -> Generator[MutableMapping[str, Any], None, None]:
+        for line in data:
+            try:
+                yield json.loads(line.decode(self.encoding))
+            except json.JSONDecodeError:
+                logger.warning(f"Cannot decode/parse line {line} as JSON")
+                # Handle invalid JSON lines gracefully (e.g., log and skip)
+                pass
+
+
+@dataclass
+class CsvParser(Parser):
+    # TODO: add more parameters: see read_csv for more details, e.g.: quotechar, headers,
```

Comment:
nit: I think the `TODO` should be `migrate implementation to re-use file-base classes`